### PR TITLE
Added Base functionality for WebSocket Client

### PR DESCRIPTION
**Pull request by seryal**

### CODE REVIEW

**Comment by pascaldragon** [comment]
Some general remarks:

- due to the integration of the other icons you'll have to redo the “Cosmetic changes” commit
- please use uppercase type identifiers (`String`, `Boolean`, etc.)
- please use a lowercase `a` as prefix for parameters
- please use a lowercase `f` as prefix for fields

Also should you maybe recreate the socket if the URL is changed?

**Comment by seryal** [comment]
Yes I will do it. Do you have a complete style guide so I can stick to it?

> Also should you maybe recreate the socket if the URL is changed?

And yes you right, i fix this.

**Comment by pascaldragon** [comment]
> Yes I will do it. Do you have a complete style guide so I can stick to it?

Not yet, sorry. 😕

**Comment by seryal** [comment]
Well then, you'll have to correct me sometimes. :)

**Comment by seryal** [comment]
Done. You can check.

**Comment by pascaldragon** [comment]
I've tried the component, but I couldn't get it to work. I tried `wss://echo.websocket.org` as URL which works on [this site](https://www.websocket.org/echo.html), but I didn't get any event. 

Following additional remarks:
- you should probably check inside `Send` whether a socket exists and raise a suitable exception otherwise
- adding an `Active` property might make sense, though you'll need to check for `csLoading in ComponentState` and initiate the socket only when `Loaded` is called
- can `OnError` provide additional information? (I don't know what the websocket API provides)
- what about the other data types that can be send/received?

**Comment by seryal** [comment]
Hmm. It's simple code.
image

```pascal
procedure TWForm1.WButton1Click(Sender: TObject);
begin
  WWebSocketClient1.Url := WEdit1.Text;
  WWebSocketClient1.Connect;
end;

procedure TWForm1.WButton2Click(Sender: TObject);
begin
  WWebSocketClient1.Send(WEdit2.Text);
end;

procedure TWForm1.WWebSocketClient1Close(Sender: TObject);
begin
  Writeln('Close');
end;

procedure TWForm1.WWebSocketClient1Error(Sender: TObject);
begin
  Writeln('Error');
end;

procedure TWForm1.WWebSocketClient1Message(Sender: TObject; Data: string);
begin
  Writeln(Data);
end;

procedure TWForm1.WWebSocketClient1Open(Sender: TObject);
begin
  Writeln('Open');
end;                 
```

**Comment by seryal** [comment]
> * you should probably check inside `Send` whether a socket exists and raise a suitable exception otherwise

Yes. It's  true. Ok.
> * adding an `Active` property might make sense, though you'll need to check for `csLoading in ComponentState` and initiate the socket only when `Loaded` is called

I use **`Connect`** method for initiate connect. But property **`Active`**, yes, it will be more correct by analogy with many components.
> * can `OnError` provide additional information? (I don't know what the websocket API provides)

I will study this question.
> * what about the other data types that can be send/received?

And will study this too.

**Comment by pascaldragon** [comment]
I found out what my problem was: I used `WMemo1.Lines.Add('xyz')` and that didn't trigger a change of the memo. I have fixed that (for now) and with that I can see that your component is indeed working correctly.
So if you address the above points and also add the component to the `RegisterComponentRequirements`, I see no problem to integrate it. 😊

**Comment by seryal** [comment]
> I found out what my problem was: I used `WMemo1.Lines.Add('xyz')` and that didn't trigger a change of the memo. I have fixed that (for now) and with that I can see that your component is indeed working correctly.
> So if you address the above points and also add the component to the `RegisterComponentRequirements`, I see no problem to integrate it. 😊

Nice. I will do it, but tomorrow. Now i go sleep 03:54 AM (+07 UTC) 😊

**Comment by seryal** [comment]
> Following additional remarks:
> 
> * you should probably check inside `Send` whether a socket exists and raise a suitable exception otherwise

Fixed.
> * adding an `Active` property might make sense, though you'll need to check for `csLoading in ComponentState` and initiate the socket only when `Loaded` is called

I tried create property `Active`, but all actions happen asynchronously and the `Active` property is set only when a connection is established, i see a problem in this. Therefore, i left the `Connect` method and the methods `Close(...)`, and also the property `Connected` for check connection state. 
> * can `OnError` provide additional information? (I don't know what the websocket API provides)

`OnClose` event provide additional information, this event fire when error too.
And i have question, https://datatracker.ietf.org/doc/html/rfc6455 11.7 paragraph have description about `Close` codes.
In which source file can text constants be defined?
> * what about the other data types that can be send/received?

WebSocket provide text data and binary data but it is still necessary to study.
At least in WEB applications, text messages are used more often (JSON strings). I think at the first stage this functionality will be enough.

**Comment by pascaldragon** [comment]
> > Following additional remarks:
> > 
> > * you should probably check inside `Send` whether a socket exists and raise a suitable exception otherwise
> 
> Fixed.
>

Thanks.

> > * adding an `Active` property might make sense, though you'll need to check for `csLoading in ComponentState` and initiate the socket only when `Loaded` is called
> 
> I tried create property `Active`, but all actions happen asynchronously and the `Active` property is set only when a connection is established, i see a problem in this. Therefore, i left the `Connect` method and the methods `Close(...)`, and also the property `Connected` for check connection state.
> 

Fair enough.

> > * can `OnError` provide additional information? (I don't know what the websocket API provides)
> 
> `OnClose` event provide additional information, this event fire when error too.
> And i have question, https://datatracker.ietf.org/doc/html/rfc6455 11.7 paragraph have description about `Close` codes.
> In which source file can text constants be defined?
> 

You can use `WCLStrConsts` for those. 😊

> > * what about the other data types that can be send/received?
> 
> WebSocket provide text data and binary data but it is still necessary to study.
> At least in WEB applications, text messages are used more often (JSON strings). I think at the first stage this functionality will be enough.

What if server replies with non-string data? Just so that this is handled gracefully...

File: widgets/extctrls.pas [hunk 1]
```diff
@@ -190,7 +190,11 @@ implementation
 { TCustomWebSocketClient }
 
 function TCustomWebSocketClient.WebSocketMessageHandler(aEvent: TEventListenerEvent): Boolean;
+var
+  s: TJSMessageEvent;
 begin
+  s := TJSMessageEvent(aEvent);
+  writeln(s.data);  // <- how get type? for text -
```
Comment (seryal) [inline widgets/extctrls.pas]: I have problem with this place. 
I cannot get the type of message.
For text message in debuger i see:
image

For binary message in debuger i see
image

`aEvent._type`  always = "message" for string and binary message

Comment (seryal) [reply widgets/extctrls.pas]: I think found :)
`GetValueType(s.Data)`

**Comment by seryal** [comment]
Finished with comments.

**Comment by pascaldragon** [comment]
Would you please redo the registration of the component so that I can easily merge this from here? 😊

**Comment by seryal** [comment]
Done.

**Comment by pascaldragon** [comment]
Thank you very much! 😊

**Comment by seryal** [comment]
I am glad to help.

**Pull request merged**